### DISCUSSION TITLE
`apk`/`apt-get` best practices

### DISCUSSION THREAD

**Discussion by GarrisonD** [mamba-org/micromamba-docker#319] [Q&A] [answered]
Here is what I am doing now and it seems working:

```Dockerfile
FROM mambaorg/micromamba:1.4.4-alpine@sha256:7bc1b7179496b7f10596ba8fce7ffd8d323566890b9f515bf1c3dd2cca1c627e

# Required for `apk add ...`
USER root

# Required for `pip install -e .`
RUN apk add --no-cache git

WORKDIR /app

COPY --chown=$MAMBA_USER:$MAMBA_USER environment.linux.lock.txt .
RUN micromamba install -y -n base -f environment.linux.lock.txt && micromamba clean --all --yes

ARG MAMBA_DOCKERFILE_ACTIVATE=1

COPY --chown=$MAMBA_USER:$MAMBA_USER . .
RUN pip install --no-cache-dir -r requirements.txt

# Discards `USER root` above
USER $MAMBA_USER
```

Would it be better to do:

```Dockerfile
# Required for `apk add ...`
USER root

# Required for `pip install -e .`
RUN apk add --no-cache git

# Discards `USER root` above
USER $MAMBA_USER
```

and only then run `micromamba install ...` and everything else?

What is the best practice of using `apk`/`apt-get` with `micromamba`?

It would be cool to be able to run just `apk`/`apt-get` without switching a user but looks like it's impossible.

**Comment by maresb** [comment, marked as answer]
It depends on who you want the owner of the `micromamba install`ed files to be. I usually switch back to `MAMBA_USER` as soon as possible so that `MAMBA_USER` owns everything.

You can also consider installing Git with `micromamba`; it is on conda-forge after all :wink:

**Reply by GarrisonD** [reply]
I tried to install `git` with `micromamba` but that takes noticeably more time and space 😞 But the idea is great!

**Reply by GarrisonD** [reply]
That would be cool if `Dockerfile` allowed something like `RUN --as=root ...` 🙂

**Reply by maresb** [reply]
> I tried to install `git` with `micromamba` but that takes noticeably more time and space disappointed But the idea is great!

Hmm, Alpine is indeed quite well-optimized for spacetime. That makes me wonder if the conda-forge version of Git is shipping unnecessary stuff. Thanks for mentioning this.

Wow, this is crazy! According to `du`, `apk add --no-cache git` adds 9.4MB, while `micromamba install -y -n base -c conda-forge git && micromamba clean --all --yes` adds 296.4 MB.
